design-proposals/macvtap-mode-for-pod-networking/macvtap-mode-for-pod-networking.md

```diff
@@ -0,0 +1,166 @@
+
+**Macvtap mode**
+
+There is **macvtap-cni** project, which allows you to bind physical device directly to the VM:
+
+```
+{Ethernet}---[eth0]
+           ^
+           |
+           '------->[macvtap]<--->[eth0]
+         ^--host--^ ^---pod---^ ^--vm--^
+```
+
+
+However it does not work with pod networking, which is always based on advanced logic and veth interfaces.
+But **having a pod network is still mandatory for the many environments where VM is communicating with pods and Kubernetes services.**
+
+This change allows to bind the macvtap interface directly to veth, so the VM will get performance of native pod networking:
+
+```
+{CNI magic}  [veth]<--->[veth]
+                     ^
+                     |
+                     '-->[macvtap]<---->[eth0]
+^---------host--------^ ^----pod---^  ^--vm--^
+```
+
+As shown by benchmarks, it gets almost the same throughput as native veth interface:
+
+![flannel](latency-flannel.svg)
+![cilium](latency-cilium.svg)
+
+## Goals
+Provide a more efficient method to bind pod networking.
+
+## Non Goals
+Same as bridge the macvtap mode for pod networking does not support live-migration, unless the [`NetworkAwareLiveMigration`](https://github.com/kubevirt/community/pull/182) feature be merged.
+
+## Definition of Users
+Users who are looking for less latency mode to bind pod networking.
+
+## User Stories
+* As a user / admin, I want to bind a VM directly to a pod network without having overhead on additional network interfaces and masquerade.
+
+## Repos
+- [KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+KubeVirt already has macvtap binding mode, but it is working only with [macvtap-cni](https://github.com/kubevirt/macvtap-cni).  
+This change is going to extend its logic to make it working with pod network and other CNIs as well.
+
+If `macvtap` mode requested, then the folowing check is invoked:
+
+- If the pod interface is **macvtap**, usually provided by macvtap-cni, then attach VM directly to it. This is left for compatibility with macvtap-cni.
+
+- If the pod interface is **veth** or any other type, then create a macvtap interface and bind to it.  
+  Due to the inability to use a downward veth interface for communicating with VM, you also have to prepare macvlan just to bind DHCP server to it.
+
+  The final scheme:  
+  ![](macvtap-mode-for-pod-networking.svg)
+
+Also this change extends existing macvtap driver to write cached information about the interface and prepare DHCP config.
+
+**How is the macvtap device created?**
+
+The new flag is added to the **tap-device-maker** binary. If it is called with `--parent-name` specified, then macvtap device will be created instead of tap.  
+But unlike tap, before running VM with macvtap interface, you need need to:
+
+- Set device rule for the cgroup running this pod:
+
+  *more details here: https://www.kernel.org/doc/Documentation/cgroup-v1/devices.txt*
```

Comment:
Will you have to write different device rules for cgroups-v1 / v2 ?

Reply:
No, I just reused existing cgroup manager interface:
https://github.com/kubevirt/kubevirt/blob/1daa0f3219aaa69891383b9c1d0158ae4fe15e91/pkg/network/driver/common.go#L446-L467

It should be agnostic to cgroups version:
https://github.com/kubevirt/kubevirt/blob/3bee69d3e9ce9cf85e97ff969a07eb29cf328d04/pkg/virt-handler/cgroup/cgroup.go#L42-L46

Meanwhile we use the same logic for hot-plugging devices:
https://github.com/kubevirt/kubevirt/blob/417fce4b5643e341a91d2fbbd2c316c2e9ca4ae6/pkg/virt-handler/hotplug-disk/mount.go#L490-L500